khal: standardize derivation

doronbehar opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

A few days ago, when updating all of my packages, I encountered build errors for khal / vdirsyncer mostly due to failing tests. When quickly patching an overlay to fix that, I noticed a few bad things in khal's `default.nix`:

1. `pkgs` is an argument to the derivation. This is not good because:
    - It's harder to change an input used inside `khal` such as `pkgs.vdirsyncer`. You have to use:
```nix
khal = super.khal.override {
  pkgs = pkgs // {
    vdirsyncer = vdirsyncer.override { ... };
  };
}
```
   * I _think_ that If you override another derivation via another overlay, that doesn't apply to what passed to that `pkgs` argument.
2. There's a `with python3.pkgs;` at the top - which makes it hard to derive python deps and non python deps (readability).
3. `pkgs.vdirsyncer` was used in `propagatedBuildInputs`. This is bad because:
   -  it is required only for tests and although used frequently together along with khal, it isn't strictly required by it to run.
   - It increases closure size if somebody uses `vdirsyncerStable`.

Now, on master (without this PR) the tests don't fail but I strongly think this change is beneficial, for general maintainability and conformism with other unwritten conventions in Nixpkgs.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

doronbehar commented:
@GrahamcOfBorg build

veprbl commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20288,7 +20288,7 @@ in
   kgraphviewer = libsForQt5.callPackage ../applications/graphics/kgraphviewer { };
 
-  khal = callPackage ../applications/misc/khal { };
+  khal = python3.pkgs.callPackage ../applications/misc/khal { };
```
How can I override this with python38?

doronbehar replied:
You are right. Considering this and your other comment, I concluded to take this 1 more step further. The next set of commits follow https://nixos.org/nixpkgs/manual/#python and use `toPythonModule` and use `callPackage` from `python-packages.nix`. I did this for all utilities we package made by `@pimutils`. This includes khard, khal, todoman and vdirsyncer.

As for your particular question, now you should be able to use `python38Packages.khal`. You can do this for all other utilities.

veprbl commented:
> There's a with python3.pkgs; at the top - which makes it hard to derive python deps and non python deps (readability)

How this improves the situation?

doronbehar committed:
vdirsyncer: standardize derivation

Use toPythonApplication and use callPackage from python-packages.nix.
Make vdirsyncerStable somewhat functional again, but mark it as broken
for Python 3.6 or higher.

veprbl commented on pkgs/top-level/python-packages.nix:
```diff
@@ -859,6 +859,23 @@ in {
 
   kconfiglib = callPackage ../development/python-modules/kconfiglib { };
 
+  todoman = disabledIf (!isPy3k) callPackage ../applications/office/todoman {
```
IIRC:

* `python-packages` can't include anything outside of pkgs/development/python-modules
* `pythonPackages` is for python modules, not for applications

doronbehar replied:
OK so how will I be able to write this in a way that satisfies:

- vdirsyncer is available as a module in `python3.pkgs` so `khal` will be able to use it from the same python packages set used to build it while it's also available as an application.
- The python version should be override-able for all of these utilities.

I'm considering using [toPythonPackage](https://nixos.org/nixpkgs/manual/#topythonmodule-function) instead but it seems it's only appropriate for bindings where in the case of `vdirsyncer` it's an almost pure python library.

FRidh replied:
`toPythonApplication`

doronbehar replied:
OK. In the latest changes I used `toPythonPacakge` only for `vdirsyncer` which is imported as a python module in `khal`'s tests. @veprbl what's all packages by `@pimutils` considered an application, are override-able with `python3 = python38;` and specifically vdirsyncer with `python38` can be used via: `python38.pkgs.vdirsyncer`.

FRidh reviewed (changes requested):
If they need to be used as libraries, add them there. If they're also used as applications, use `toPythonApplication` in `all-packages.nix`.

doronbehar committed:
khal: format inputs and arguments

doronbehar committed:
khal: clean up deps and use with python3.pkgs; only when needed

doronbehar committed:
todoman: format inputs and arguments

doronbehar committed:
todoman: remove glib locale fixes

Don't know why they were there in the first place :/.

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1585,6 +1585,13 @@ in {
+    inherit (pkgs) pkg-config openssl rustPlatform;
+  };
+
+  vdirsyncerStable = disabledIf (!isPy3k) callPackage ../tools/misc/vdirsyncer/stable.nix { };
```
only *libraries* should be imported, not applications. Either move `vdirsyncer` under `python-modules` and make it use `buildPythonPackage`, or use `toPythonModule`. But please avoid that with pure Python packages because it would not make much sense. Here I argue moving `vdirsyncer` is the preferred solution.

doronbehar replied:
I'm not sure I understand what you mean "moving" because it's already in python-packages.nix, but I agree that using `buildPythonPackage` is more appropriate now.

FRidh replied:
> Either move vdirsyncer under `python-modules`

`python-modules` is the folder, so this is about the the actual expression.

doronbehar replied:
I see, done.

doronbehar committed:
vdirsyncer: use buildPythonPackage as it's a package now

doronbehar committed:
vdirsyncer: move to python-modules/

doronbehar commented:
BTW @FRidh I encountered that `overrideAttrs` issue while encountering the issue failed tests this PR might. So please if you could take 1 more look here.

doronbehar commented:
@FRidh I know it's been a long time but this is still bothering me. Also, today I noticed `todoman` references `glibLocales` which is ~210Mb in size. Commit https://github.com/NixOS/nixpkgs/pull/85993/commits/2260aed8ada148975cca987c53e801c3514b998c in this PR fixes this.

doronbehar commented:
I'm closing in favor of smaller PRs which hopefully will be easier to review.
